pallets/block-rewards-hybrid/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+
+    #[pallet::genesis_config]
+    #[cfg_attr(feature = "std", derive(Default))]
+    pub struct GenesisConfig {
+        pub reward_config: RewardDistributionConfig,
+    }
+
+    #[pallet::genesis_build]
+    impl<T: Config> GenesisBuild<T> for GenesisConfig {
+        fn build(&self) {
+            assert!(self.reward_config.is_consistent());
+            RewardDistributionConfigStorage::<T>::put(self.reward_config.clone())
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Sets the reward distribution configuration parameters which will be used from next block reward distribution.
+        ///
+        /// It is mandatory that all components of configuration sum up to one whole (**100%**),
+        /// otherwise an error `InvalidDistributionConfiguration` will be raised.
+        ///
+        /// - `reward_distro_params` - reward distribution params
+        ///
+        /// Emits `DistributionConfigurationChanged` with config embeded into event itself.
+        ///
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::set_configuration())]
+        pub fn set_configuration(
+            origin: OriginFor<T>,
+            reward_distro_params: RewardDistributionConfig,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+
+            ensure!(
+                reward_distro_params.is_consistent(),
+                Error::<T>::InvalidDistributionConfiguration
+            );
+            RewardDistributionConfigStorage::<T>::put(reward_distro_params.clone());
+
+            Self::deposit_event(Event::<T>::DistributionConfigurationChanged(
+                reward_distro_params,
+            ));
+
+            Ok(().into())
+        }
+    }
+
+    impl<Moment, T: Config> OnTimestampSet<Moment> for Pallet<T> {
+        fn on_timestamp_set(_moment: Moment) {
+            let rewards = Self::calculate_rewards(T::MaxBlockRewardAmount::get());
+            let inflation = T::Currency::issue(rewards.sum());
+            Self::distribute_rewards(inflation, rewards);
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Calculates the amount of rewards for each beneficiary
+        ///
+        /// # Arguments
+        /// * `block_reward` - the block reward amount
+        ///
+        fn calculate_rewards(block_reward: Balance) -> Rewards {
+            let distro_params = Self::reward_config();
+
+            // Pre-calculate balance which will be deposited for each beneficiary
+            let base_staker_balance = distro_params.base_staker_percent * block_reward;
+            let dapps_reward = distro_params.dapps_percent * block_reward;
+            let collators_reward = distro_params.collators_percent * block_reward;
+            let treasury_reward = distro_params.treasury_percent * block_reward;
+
+            // This is part is the TVL dependant staker reward
+            let adjustable_balance = distro_params.adjustable_percent * block_reward;
```

Comment:
Might be a good idea to add one `debug_assert` here and ensure that the sum of all equals the block reward.

Reply:
It will no be equal actually if we are below `ideal_dapps_staking_tvl`